cachi2/core/package_managers/generic/main.py

```diff
@@ -47,8 +53,47 @@ def _resolve_generic_lockfile(source_dir: RootedPath, output_dir: RootedPath) ->
 
     log.info(f"Reading generic lockfile: {lockfile_path}")
     lockfile = _load_lockfile(lockfile_path)
-    for artifact in lockfile.artifacts:
+    artifacts: List[GenericArtifact] = [
+        GenericArtifact.from_lockfile_artifact(a) for a in lockfile.artifacts
+    ]
+
+    # output_dir is now the root and cannot be escaped
+    output_dir = output_dir.re_root(DEFAULT_DEPS_DIR)
+
+    for artifact in artifacts:
         log.debug(f"Resolving artifact: {artifact.download_url}")
+        # make sure target does not point outside the output directory
+        try:
+            destination = output_dir.join_within_root(artifact.resolved_target)
+        except PathOutsideRoot as e:
+            e.solution = (
+                f"Artifact '{artifact.download_url}' has target which falls "
+                "outside of the dependencies root. Refusing to proceed."
+            )
+            raise
+
+        # make sure that targets do not overlap
+        if any([a.has_same_resolved_path(destination) for a in artifacts]):
```

Comment:
I have a small suggestion to avoid needing this nested loop:
```
existing_paths = set()

for artifact in artifacts:
   destination = output_dir.join_within_root(artifact.resolved_target)
   
   if destination in existing_paths:
      raise Exception()
   else:
       existing_paths.add(destination)     
```

And ideally keep the result of `destination` to avoid having to recalculate it during the download phase. I think this check could fit well into a tiny utility function.

Reply:
I used the general idea of this approach. Destination is now `_resolved_path` in the `LockfileArtifact`